torchaudio/datasets/utils.py

```diff
@@ -76,41 +65,85 @@ def makedir_exist_ok(dirpath):
+    """Execute the correct download operation.
+    Depending on the size of the file online and offline, resume the
+    download if the file offline is smaller than online.
 
     Args:
-        url (str): URL to download file from
-        root (str): Directory to place downloaded file in
-        filename (str, optional): Name to save the file under. If None, use the basename of the URL
-        md5 (str, optional): MD5 checksum of the download. If None, do not check
+        url (str): Url.
+        download_folder (str): Folder to download file.
+        hash_value (str): Hash for url.
+        hash_type (str): Hash type.
     """
-    from six.moves import urllib
 
-    root = os.path.expanduser(root)
-    if not filename:
-        filename = os.path.basename(url)
-    fpath = os.path.join(root, filename)
+    filepath = os.path.join(download_folder, os.path.basename(url))
 
-    makedir_exist_ok(root)
+    req = urllib.request.Request(url)
+    if os.path.exists(filepath):
+        mode = "ab"
+        local_size = os.path.getsize(filepath)
 
-    # downloads file
-    if os.path.isfile(fpath):
-        print("Using downloaded file: " + fpath)
+        # If the file exists, then download only the remainder
+        req.headers["Range"] = "bytes={}-".format(local_size)
     else:
-        try:
-            print("Downloading " + url + " to " + fpath)
-            urllib.request.urlretrieve(url, fpath, reporthook=gen_bar_updater())
-        except (urllib.error.URLError, IOError) as e:
-            if url[:5] == "https":
-                url = url.replace("https:", "http:")
-                print(
-                    "Failed download. Trying https -> http instead."
-                    " Downloading " + url + " to " + fpath
+        mode = "wb"
+        local_size = 0
+
+    # If we already have the whole file, there is no need to download it again
+    url_size = int(urllib.request.urlopen(url).info().get("Content-Length", -1))
+    if url_size == local_size:
+        if hash_value and not validate_download_url(filepath, hash_value, hash_type):
+            raise RuntimeError(
+                "The hash of {} does not match. Delete the file manually and retry.".format(
+                    filepath
                 )
-                urllib.request.urlretrieve(url, fpath, reporthook=gen_bar_updater())
-            else:
-                raise e
+            )
+
+        return
+
+    with open(filepath, mode) as fpointer, urllib.request.urlopen(
+        req
+    ) as upointer, tqdm(
+        unit="B", unit_scale=True, unit_divisor=1024, total=url_size
+    ) as pbar:
+
+        num_bytes = 0
+        block_size = 32 * 1024
+        while True:
+            chunk = upointer.read(block_size)
+            if not chunk:
+                break
+            fpointer.write(chunk)
+            num_bytes += len(chunk)
+            pbar.update(len(chunk))
+
+
+def validate_download_url(filepath, hash_value, hash_type="sha256"):
```

Comment:
This works with any file right? So it could just be called "validate" and take a path. Even better, take a buffer.

Reply:
So you want to suggest removing the `with open(filepath, "rb")` statement?

Reply:
Yes and instead accept a file_like object.

Reply:
Sure, I'm ok with that.